cpp/ql/test/query-tests/Critical/UnsafeUseOfThis/test.cpp

```diff
@@ -0,0 +1,50 @@
+struct A { virtual void f() = 0; };
+
+struct B;
+void call_f(B*);
+
+struct B : public A {
+	B() {
+		call_f(this);
+	}
+
+    B(B& b) {
+        b.f(); // BAD: undefined behavior
+    }
+
+	~B() {
+		f(); // BAD: undefined behavior
+	}
+};
+
+struct C : public B {
+	C(bool b) {
+		call_f(this);
+
+		if(b) {
+			this->f(); // GOOD: Not a 'must' flow
```

Comment:
I don't understand why this is `GOOD`. Won't the program crash in practice if this line of code is ever executed?

Reply:
Maybe it's not the right way to annotate this, no. The program will indeed crash if we reach it. We don't flag it due to the post-domination criteria. Initially, I inserted the post-dominance criteria to focus the query on problems in a constructor's member initialization list, but maybe we actually want to flag this one as well (and so the post-dominance criteria should be removed). With post-dominance removed, we still won't flag something like this, though:

```cpp
C(bool b) {
  B* that = nullptr;
  if(b) { that = this; }
  that->f();
}
```
since this relies on 'may' flow (i.e., flow through `Phi`s).

Reply:
It sounds alright to me that we're not flagging that `nullptr` example for now.

While it's easy to argue that `C(bool b)` should not crash regardless of the value of `b`, that doesn't generalise. For `C(bool b1, bool b2)`, should we give an alert if the constructor crashes on just one of the four combinations of `b1` and `b2`? For programmers who don't like to code defensively, the answer might be no. Even for defensive programmers, they might have written `assert(b1 || b2);` in the first line of the constructor and decided that this was enough.

If we are to keep the post-dominance criterion, I'd like to see a test that justifies it. A line of code that crashes if it's ever executed (and isn't dead code) should definitely have an alert on it.

Reply:
I've removed the post-dominance requirement in 0a6a22562b5166cc53398625a271a5084e41214a.

Reply:
I hope you'll make a separate PR to add those post-dominance predicates in `IRBlock` anyway. Someone will need them sooner or later, and they had good QLDoc on them!

Reply:
I have them in a local branch. I can create a PR for them later today.

Reply:
Done in https://github.com/github/codeql/pull/4673.